teps/0065-retry-failed-tasks-on-demand.md

```diff
@@ -0,0 +1,262 @@
+
+Ability to `retry` failed tasks is especially useful, where `tekton` is a
+backend for running Machine learning pipelines. A machine learning pipeline
+may consist of tasks moving large amount of data and then training ml models,
+all of it can be very resource consuming and inability to retry would require
+a user to start the entire pipeline over. Sometimes, the failure could be due
+to temporary service outages. For example, after training the model, a task
+reporting the metrics fails due to temporary service outage. A retry after
+some time could easily fix it.
+
+A pipeline may be defined with various tasks, and some tasks might move a
+large amount of data and incur cost. This `retry` mechanism has substantial
+value, where each task of the pipeline incurs a significant computing resources, 
+e.g. `tekton` is used as a backend for ML pipelines.
+
+_Why do we need a new `retry` mechanism when we already support retry in 
+`Pipeline` tasks?_ 
+
+The present `retry` field can only be defined at the time of creation of
+pipeline. This is not suitable for use cases, where a manual intervention
+is necessary to decide whether a rerun is required or not.
+For example, if a service outage is causing a particular task failure, then
+retrying `n` times, won't help, unless we wait for the service to be back
+again and retry. For such manual interventions, we need on-demand `retry`
+mechanism.
+
+Another concocted example, if `Pipeline` were to represent a  CI/CD job, then
+tasks represent test suit, stress test and benchmarks. Now, we need a way to
+know whether a failure was due to some regression, or it is due to flakiness
+of jobs itself or temporary service outage. In this case, simply retrying `n`
+number of times does not seem to help with optimal resource consumption.
+
+### Goals
+
+1. Explore both the merits and demerits in having a new mechanism for on-demand
+   retrying, an _only a failed_ pipeline.
+2. A pipeline may either have failed due to some failures in the tasks or may
+   be user invoked cancel request. Retry only the failed/canceled tasks for a
+   failed `pipelineRun`.
+
+### Non-Goals
+
+1. Retry of successful pipeline runs or anything other than a failed pipeline/task
+   run.
+2. Changing existing retry mechanism.
+3. Manage checkpointing of pipeline state or workspaces, etc. A `pipelineRun`'s
+   state stored in etcd is used as is.
+4. Determine, a failed tasks dependencies i.e. figuring out what
+   all dependent tasks are needed to rerun the failed task.
+   
+### Use Cases (optional)
+
+1. `PipelineRun` can be very resource consuming, and are sometimes susceptible to
+   fail due to transient conditions. For example, due to service outage of a 
+   particular service. In such cases, it is not enough to be retried `n` times,
+   a manual invocation of retry is required.
+   
+2. It will be possible to cancel (e.g. preemption) any running `PipelineRun`, and
+   resume at a later point.
+
```

Comment:
> We already have the concept of rerun in Kubeflow pipeline where we are running the pipeline again with a new pipelineRun. One of the main problems we see is that some users might use pipelinerun.uid and pipelinerun name to distinguish their jobs. So we want a retry feature that can keep these Tekton context variables the same (like pipelinerun name and uid) when users retry the same job.

Thanks for explaining this @Tomcli - I think this could be a useful addition to the use cases section

Reply:
@Tomcli @ScrapCodes  I think this is a really difficult requirement to satisfy - one might argue that if a system builds an abstraction on top of Tekton, that abstraction might have it's own IDs and those should be the one users can refer to. 

Tekton has a strong association between the `PipelineRun` / `TaskRun` and the *single* execution of a `Pipeline` / `Task` respectively. The status of the resources and metrics are modelled under the assumption that the resource is executed once. Even in the case of pipeline task retries that exists today, a new `TaskRun` is created for each execution.

If we started having a second execution associated to the same ID, we would need to remodel the status resource, since we would need to be able to store an array of statuses. We would also need to consider what the implications are in terms of execution history (tekton results) and supply chain security (tekton chains).

An alternative solution might be to create a new `PipelineRun` and set a specific label that marks the new `PipelineRun` as being part of a set of retries of the original one. For instance we could have `tekton.dev/pipelineRunSeriesID: <uuid-of-the-first-pipeline>` set on all pipelines (even the original execution).
Users, as well as tools (tkn, dashboard) could use that label to collate the retries together.

Yet another alternative could be to have a new CRD, which represents a group of pipelines, i.e. the original one and its retries. but again users would have to know they need to look for that resource instead of a pipelinerun.